pkg/test/framework/components/echo/util/traffic/generator.go

```diff
@@ -0,0 +1,105 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package traffic
+
+import (
+	"fmt"
+	"time"
+
+	"istio.io/istio/pkg/test/framework/components/echo"
+)
+
+const (
+	waitTimeout     = 30 * time.Second
+	defaultInterval = 1 * time.Second
+)
+
+// Config for a traffic Generator.
+type Config struct {
+	// Source of the traffic.
+	Source echo.Instance
+
+	// Options for generating traffic from the Source to the target.
+	Options echo.CallOptions
+
+	// Interval between successive call operations. If not set, defaults to 1 second.
+	Interval time.Duration
+}
+
+// Generator of traffic between echo instances.
+type Generator interface {
+	// Run the Generator and start sending traffic.
+	Run(stop chan struct{})
+
+	// WaitForResult waits for a short while for the Generator to complete
+	// and then returns the Result.
+	WaitForResult() (Result, error)
+}
+
+// NewGenerator returns a new Generator with the given configuration.
+func NewGenerator(cfg Config) Generator {
+	fillInDefaults(&cfg)
+	return &generator{
+		Config:  cfg,
+		stopped: make(chan struct{}),
+	}
+}
+
+var _ Generator = &generator{}
+
+type generator struct {
+	Config
+	result  Result
+	stopped chan struct{}
+}
+
+func (g *generator) Run(stop chan struct{}) {
+	for {
+		t := time.NewTimer(g.Interval)
+		select {
+		case <-stop:
+			t.Stop()
+			close(g.stopped)
+			return
+		case <-t.C:
+			g.result.Add(g.Source.Call(g.Options))
+			t.Reset(g.Interval)
+		}
+	}
+}
+
+func (g *generator) WaitForResult() (Result, error) {
+	t := time.NewTimer(waitTimeout)
+	select {
+	case <-g.stopped:
+		t.Stop()
+		return g.result, nil
+	case <-t.C:
+		return Result{}, fmt.Errorf("timed out waiting for result")
+	}
+}
+
+func fillInDefaults(cfg *Config) {
+	// Each call must be a single request
+	cfg.Options.Count = 1
+
+	if cfg.Interval == 0 {
+		cfg.Interval = defaultInterval
+	}
+
+	if cfg.Options.Validator == nil {
+		cfg.Options.Validator = echo.ExpectOK()
```

Comment:
unrelated to this PR: if this isn't the default already in the main code, it probably should be? otherwise people may assume its checked when its not